TCC.Lib/Dependencies/ExternalDependecies.cs

```diff
@@ -0,0 +1,142 @@
+
+        public string Root => Path.GetDirectoryName(typeof(ExternalDependecies).Assembly.CodeBase.Replace("file:///", ""));
+
+        public string GetPath(ExternalDep dependency)
+        {
+            string exePath = Path.Combine(Root, dependency.ExtractFolder, dependency.ExeName);
+            if (!File.Exists(exePath))
+            {
+                throw new FileNotFoundException(dependency.Name + " not found in " + exePath);
+            }
+            return exePath;
+        }
+
+        public async Task<string> EnsureDependency(ExternalDep dependency, IProgress<int> percent)
+        {
+            string path = Path.Combine(Root, dependency.ExtractFolder);
+            string exePath = Path.Combine(Root, dependency.ExtractFolder, dependency.ExeName);
+            string download_lock = Path.Combine(Root, dependency.Name + ".lock");
+            if (!Directory.Exists(path) || !File.Exists(exePath))
+            {
+                using (new FileStream(download_lock, FileMode.OpenOrCreate, FileAccess.Read, FileShare.None))
+                {
+                    if (File.Exists(exePath))
+                    {
+                        percent.Report(100);
+                        return exePath;
+                    }
+
+                    string target = Path.Combine(path, string.IsNullOrEmpty(dependency.ZipFilename)
+                        ? dependency.ExeName
+                        : dependency.ZipFilename);
+
+                    Directory.CreateDirectory(path);
+
+                    using (var wc = new WebClient())
+                    {
+                        wc.DownloadProgressChanged += (s, e) =>
+                        {
+                            percent.Report((int)(e.BytesReceived / e.TotalBytesToReceive * 100));
+                        };
+                        wc.DownloadFileCompleted += (s, e) =>
+                        {
+                            percent.Report(100);
+                        };
+                        await wc.DownloadFileTaskAsync(dependency.URL, target);
+                    }
+
+                    if (!string.IsNullOrEmpty(dependency.ZipFilename))
+                    {
+                        ZipFile.ExtractToDirectory(target, path);
+                    }
+
+                    if (!File.Exists(exePath))
+                    {
+                        throw new FileNotFoundException(dependency.Name + " not found in " + exePath);
+                    }
+                }
+                File.Delete(download_lock);
+            }
+            percent.Report(100);
+            return exePath;
+        }
+
+        private ExternalDep _lz4 = new ExternalDep()
+        {
+            Name = "Lz4",
+            URL = @"https://github.com/lz4/lz4/releases/download/v1.8.0/lz4_v1_8_0_win64.zip",
+            ZipFilename = "lz4_v1_8_0_win64.zip",
+            ExtractFolder = "lz4",
+            ExeName = "lz4.exe"
+        };
+
+        private ExternalDep _tar = new ExternalDep()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make '_tar' 'readonly'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove the unused private field '_tar'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)